### PR TITLE
Smaller versions of Marlowe Validator

### PR DESCRIPTION
**Pull request by nau**
Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: marlowe/src/Language/Marlowe/Scripts.hs [hunk 1]
```diff
@@ -74,7 +82,7 @@ mkMarloweStateMachineTransition params SM.State{ SM.stateData=MarloweData{..}, S
     (interval@(minSlot, maxSlot), inputs) = do
     let positiveBalances = validateBalances marloweState ||
             -- Avoid creating a too-big string literal
-            P.traceError ("Invalid contract state. " `P.appendString` "There exists an account with non positive balance")
+            traceError ("M1")
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Scripts.hs:85]: Eventually, we might want to create a markdown file that lists the trace error message abbreviations and their interpretations.

File: marlowe/src/Language/Marlowe/Scripts.hs [hunk 2]
```diff
@@ -171,6 +179,122 @@ mkValidator :: MarloweParams -> Scripts.ValidatorType MarloweStateMachine
...
+    -> Bool
+smallMarloweValidator MarloweParams{rolesCurrency, rolePayoutValidatorHash} MarloweData{..} inputs ctx@ScriptContext{scriptContextTxInfo} = do
+    let slotConfig = def :: TimeSlot.SlotConfig
+    let ownInput = case findOwnInput ctx of
+            Just i -> i
+            _      -> traceError "I0" {-"Can't find validation input"-}
+    let scriptInValue = txOutValue $ txInInfoResolved ownInput
+    let (minTime, maxTime) =
+            case txInfoValidRange scriptContextTxInfo of
+                Interval.Interval (Interval.LowerBound (Interval.Finite l) True) (Interval.UpperBound (Interval.Finite h) False) -> (l, h)
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Scripts.hs:197]: FYI: I double-checked in the [ledger spec](https://hydra.iohk.io/build/7840126/download/1/shelley-ma.pdf) that the `[closed, open)` interpretation here of the validity interval is correct.

File: marlowe/src/Language/Marlowe/Scripts.hs [hunk 3]
```diff
@@ -171,6 +179,122 @@ mkValidator :: MarloweParams -> Scripts.ValidatorType MarloweStateMachine
...
+    payoutConstraints payoutsByParty = all payoutToTxOut payoutsByParty
+      where
+        payoutToTxOut (party, value) = case party of
+            PK pk  -> traceIfFalse "P" $ value `Val.leq` valuePaidTo scriptContextTxInfo pk
+            Role role -> let
+                dataValue = Datum $ PlutusTx.toBuiltinData role
+                hsh = findDatumHash dataValue scriptContextTxInfo
+                addr = Ledger.scriptHashAddress rolePayoutValidatorHash
+                in traceIfFalse "R" $ any (checkScriptOutput addr hsh value) allOutputs
+
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Scripts.hs:295]: Why do payments to public keys test value paid using an inequality, but payments to roles use equality?

My understanding of `payoutConstraints` is
1. When paying a public key, `payout specified by contract` `<=` `actual payout in transaction`.
2. When paying a role, `payout specified by contract` `==` `payout of at least one output`.

In principle this test would permit paying more than the contract permits to a party. For a public key, the `<=` logic permits this. For a role, there could be several outputs totaling more than the specified payout so long as one output contains the exact payout value. So this constraint reads "every party gets paid at least what the contract specifies"? In practice, I presume that the logic of particular contracts might constrain against excess payments?

File: marlowe/src/Language/Marlowe/Scripts.hs [hunk 4]
```diff
@@ -171,6 +179,122 @@ mkValidator :: MarloweParams -> Scripts.ValidatorType MarloweStateMachine
...
+                    Close -> True
+                    _ -> let
+                        totalIncome = foldMap collectDeposits inputs
+                        totalPayouts = foldMap snd payoutsByParty
+                        finalBalance = inputBalance + totalIncome - totalPayouts
+                        outConstrs = OutputConstraint
+                                    { ocDatum = marloweData
+                                    , ocValue = finalBalance
+                                    }
+                        in checkOwnOutputConstraint ctx outConstrs
```
Comment (bwbush) [inline marlowe/src/Language/Marlowe/Scripts.hs:247]: FYI: [`checkOwnOutputConstraints`](https://github.com/input-output-hk/plutus-apps/blob/a6868cde87600051a4cc808777d665d0b859853c/plutus-ledger/src/Ledger/Constraints/OnChain.hs#L42-L51) contains logic related to min-ADA. At least one output must simultaneously satisfy these conditions:
1. The datum hash matches.
2. The non-ADA value exactly matches the constraint.
3. The ADA value is `>=` the constraint and `<=` 2 ADA plus the constraint.

This shouldn't adversely affect this validator.

**Review by bwbush — COMMENTED** [review]
This looks good, and I only raised one substantial question about inequality vs equality constraints on payouts.

FYI: The [bwbush/marlowe-cli](https://github.com/input-output-hk/marlowe-cardano/tree/bwbush/marlowe-cli) branch contains additional `marlowe-cli` functionality that I'll create a pull request for as soon as this pull request is merged.

**Review by palas — APPROVED** [review]
Looks great! Thanks 👍

**Pull request merged**